cpp/src/parquet/encoding.cc

```diff
@@ -2236,6 +2324,76 @@ class DeltaByteArrayDecoder : public DecoderImpl,
   ByteArray last_value_;
 };
 
+// ----------------------------------------------------------------------
+// BYTE_STREAM_SPLIT
+
+template <typename DType>
+class ByteStreamSplitDecoder : public DecoderImpl, virtual public TypedDecoder<DType> {
+ public:
+  using T = typename DType::c_type;
+  explicit ByteStreamSplitDecoder(const ColumnDescriptor* descr);
+
+  int Decode(T* buffer, int max_values) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::Accumulator* builder) override;
+
+  int DecodeArrow(int num_values, int null_count, const uint8_t* valid_bits,
+                  int64_t valid_bits_offset,
+                  typename EncodingTraits<DType>::DictAccumulator* builder) override;
+
+  void SetData(int num_values, const uint8_t* data, int len) override;
+
+ private:
+  int num_values_decoded_{0U};
+};
+
+template <typename DType>
+ByteStreamSplitDecoder<DType>::ByteStreamSplitDecoder(const ColumnDescriptor* descr)
+    : DecoderImpl(descr, Encoding::BYTE_STREAM_SPLIT) {}
+
+template <typename DType>
+void ByteStreamSplitDecoder<DType>::SetData(int num_values, const uint8_t* data,
+                                            int len) {
+  DecoderImpl::SetData(num_values, data, len);
+  num_values_decoded_ = 0;
+}
+
+template <typename DType>
+int ByteStreamSplitDecoder<DType>::Decode(T* buffer, int max_values) {
+  constexpr size_t num_streams = sizeof(T);
+  using UnsignedType = typename UnsignedTypeWithWidth<sizeof(T)>::type;
+  const int values_to_decode = std::min(num_values_ - num_values_decoded_, max_values);
+  for (int i = 0; i < values_to_decode; ++i) {
+    UnsignedType reconstructed_value_as_uint{0U};
+    for (size_t b = 0; b < num_streams; ++b) {
+      const size_t byte_index = b * num_values_ + num_values_decoded_ + i;
```

Comment:
Is it accurate/correct that this is striding by `num_values_` bytes through the encoded memory on each iteration?

Reply:
Yes, there are sizeof(T) streams and num_values_ entries per stream.
The array of F32 values [0x11223344, 0xAABBCCDD] would be encoded as
[44 DD 33 CC 22 BB 11 AA].
num_values_ is 2 here which is the stride.
So to decode the first value we go over 44, 33, 22, 11.